crates/subspace-farmer/src/single_plot_farm.rs

```diff
@@ -163,6 +163,29 @@ impl SinglePlotFarm {
                 move |&PiecesByRangeRequest { from, to }| {
                     let mut pieces_and_indexes =
                         plot.get_sequential_pieces(from, SYNC_PIECES_AT_ONCE).ok()?;
+
+                    if pieces_and_indexes.len() == 1 {
+                        let (index, mut piece) = pieces_and_indexes
+                            .into_iter()
+                            .next()
+                            .expect("Always present as length is exactly one");
+                        let index_hash = PieceIndexHash::from(index);
+                        if index_hash >= to {
+                            return Some(PiecesByRangeResponse {
+                                pieces: PiecesToPlot::default(),
+                                next_piece_index_hash: Some(index_hash),
+                            });
+                        }
+                        codec.decode(&mut piece, index).ok()?;
+                        return Some(PiecesByRangeResponse {
+                            pieces: PiecesToPlot {
+                                piece_indexes: vec![index],
+                                pieces: piece.into(),
+                            },
+                            next_piece_index_hash: None,
+                        });
+                    }
```

Comment:
If we have just 1 piece we will return it as `next_piece_index_hash` all the time

Reply:
Why? If `from >= index_hash` we should return `None`, I don't see why special case is necessary for this length.

Reply:
Oh, you right. Will revert this change then. Sorry for misleading!

Reply:
Actually, we need this special case, because if we have only one piece left which is in our range, we will always return empty `PiecesToPlot` and `next_piece_index_hash` set to this piece index hash.

Second special case needed in case when we have piece out of asked range (in this case we need to return no pieces and `next_piece_index_hash` set to `None`)

Reply:
That's why I decided to have this branch in the first place :)

Reply:
Why does generic code below need to have this bug though? If generic code below doesn't have such a bug then special case is not needed in the first place. I feel like either something is wrong or I don't get something.

Reply:
Because it didn't handle this specific case when the length of the returned array is one. Dunno how to elaborate on that.

Beforehand we tried to find some index which is out of our range, and if we didn't find one (in the case of the only one index in our range) we would just pop the last index and return it as the `next_piece_index_hash`.

Can change the generic code to workaround that, but I think current semantics is the same anyway.

Reply:
> Beforehand we tried to find some index which is out of our range, and if we didn't find one (in the case of the only one index in our range) we would just pop the last index and return it as the next_piece_index_hash.

Why would we return index that is in range as `next_piece_index_hash`? Isn't that wrong?

> Can change the generic code to workaround that, but I think current semantics is the same anyway.

I'm sorry, I don't see the problem to work around, it should return in `next_piece_index_hash` things that are out of range. If there is nothing out of range then return `None`.